Add checksumming and versions to the Translog's Checkpoint files

bleskes opened the pull request:
This prepares the infrastructure to be able to extend the checkpoint file to store more information.

bleskes commented:
@s1monw @mikemccand can you guys take a look?

PS. I would love to get rid of the `DirectOutputStreamIndexOutput` I added there, but I didn't want multiple in memory buffers - it just irked me the wrong way. It will be wonderful to hear if there is a better way in the big jungle of OutputStreams and DataOutputs

mikemccand commented on core/src/main/java/org/elasticsearch/index/translog/Checkpoint.java:
```diff
@@ -18,86 +18,135 @@
-            return new Checkpoint(new InputStreamDataInput(in));
+        try (Directory dir = new SimpleFSDirectory(path.getParent())) {
+            try (final IndexInput indexInput = dir.openInput(path.getFileName().toString(), IOContext.DEFAULT)) {
+                if (indexInput.length() == 20) {
```
Can we pull `20` out to a static constant like `OLD_CHECKSUM_FILE_LENGTH`?

mikemccand commented on core/src/main/java/org/elasticsearch/index/translog/Checkpoint.java:
```diff
@@ -18,86 +18,135 @@
+                // We checksum the entire file before we even go and parse it. If it's corrupted we barf right here.
+                CodecUtil.checksumEntireFile(indexInput);
+                final int fileVersion = CodecUtil.checkHeader(indexInput, CHECKPOINT_CODEC, INITIAL_VERSION, INITIAL_VERSION);
+                assert fileVersion == INITIAL_VERSION : "trust no one! " + fileVersion;
```
Can we make this a real check instead of `assert`?

mikemccand replied:
Oh nevermind, `CodecUtil.checkHeader` is already doing the real check.  You really do not trust anyone ;)

bleskes replied:
hehe. yeah. Left over. will remove.

mikemccand commented on core/src/main/java/org/elasticsearch/index/translog/Checkpoint.java:
```diff
@@ -18,86 +18,135 @@
         try (FileChannel channel = factory.open(checkpointFile, options)) {
-            checkpoint.write(channel);
+            Channels.writeToChannel(byteOutputStream.toByteArray(), channel);
+            // no need to force metadata, file size stays the same
```
Can you improve the comment by also saying that "... and we did the full `fsync` when we first created the file, so the directory entry doesn't change as well"?

mikemccand commented on core/src/main/java/org/elasticsearch/index/translog/Checkpoint.java:
```diff
@@ -18,86 +18,135 @@
-        out.writeLong(offset);
-        out.writeInt(numOps);
-        out.writeLong(generation);
+    static Checkpoint readNonChecksummed(DataInput in) throws IOException {
```
Maybe just a single `read` method instead of `readChecksummed` and `readNonChecksummed`?  They do the same thing?  In the future, we'll have to also pass in the version to `read` so it can read any version-specific things ...

bleskes replied:
I was going for a model where the versioning switch is not in these methods and have one method per "version". I think it's easier to clean up when code becomes old and is not needed any more.

mikemccand replied:
OK that makes sense.

mikemccand commented on core/src/main/java/org/elasticsearch/index/translog/Checkpoint.java:
```diff
@@ -109,4 +158,50 @@ public int hashCode() {
     }
+
+    /** A copy of {@link OutputStreamIndexOutput} without the intermediate buffering */
+    static class DirectOutputStreamIndexOutput extends IndexOutput {
```
Why not just use `OutputStreamIndexOutput`?  Is its buffering somehow problematic?

bleskes replied:
Because it  bugged me that we do buffering when we write to an in-memory buffer anyway. I'm not sure why that class does buffering when you can give it a buffering stream, if you want buffering? Alternatively we should allow to set the buffer size to 0 , in which cases the CRC will not be wrapped.

mikemccand commented on core/src/main/java/org/elasticsearch/index/translog/Checkpoint.java:
```diff
@@ -18,86 +18,135 @@
+        };
+        final String resourceDesc = "checkpoint(path=\"" + checkpointFile + "\", gen=" + checkpoint + ")";
+        try (final DirectOutputStreamIndexOutput indexOutput =
+                 new DirectOutputStreamIndexOutput(resourceDesc, checkpointFile.toString(), byteOutputStream)) {
```
You could maybe use `ByteArrayDataOutput` (writes directly to a fixed byte[], but since we know how many bytes we will write here that should be OK) instead?  Or `OutputStreamDataOutput`?

bleskes replied:
Oh the jungle :) `ByteArrayDataOutput` doesn't implement `OutputStream` which `(Direct)OutputStreamIndexOutput` require as a parameter. Same goes for `OutputStreamDataOutput`.

s1monw replied:
maybe its simpler to do this:

```Java
try (FileChannel channel = factory.open(checkpointFile, options)) {
  IndexOutput indexOutput = new OutputStreamIndexOutput (resourceDesc, checkpointFile.toString(), java.nio.channels.Channels.newOutputStream(channel), BUFFER_SIZE);

 // ...

}
```

bleskes replied:
I tried that too :(

Sadly IndexOutput doesn't allow you to flush it's internal buffer, so you have to close it , at which point your channel is closed, so you can't fsync it...

s1monw replied:
hmm the close method here looks like this:

```Java
@Override
  public void close() throws IOException {
    try (final OutputStream o = os) {
      // We want to make sure that os.flush() was running before close:
      // BufferedOutputStream may ignore IOExceptions while flushing on close().
      // We keep this also in Java 8, although it claims to be fixed there,
      // because there are more bugs around this! See:
      // # https://bugs.openjdk.java.net/browse/JDK-7015589
      // # https://bugs.openjdk.java.net/browse/JDK-8054565
      if (!flushedOnClose) {
        flushedOnClose = true; // set this BEFORE calling flush!
        o.flush();
      }
    }
  }
```
so I am not sure why it close the channel

bleskes replied:
because the `os` in a try with resources clause - so it's close by that.

s1monw replied:
oh yeah I missed that :/

s1monw replied:
ok I think ultimately we should just have a non-closeing FilterOutputStream to prevent this?

bleskes replied:
sadly non closing is not enough. you have to use an intermediate in memory buffer in order to make sure you write to the underlying channel in one go. If not, the crc calculation which flushes will write the first chunk and then the crc will be written as a second chunks. I had that too and guess - tests signaled it was wrong (good!)

mikemccand commented:
Thanks @bleskes, I left some comments.

bleskes commented:
@mikemccand @s1monw  I pushed another commit addressing feedback. I switched back to vanilla `OutputStreamDataOutput`, which means double buffers but it seems you guys are more comfortable with that trade off (which I agree with if there is no future coming to allow to disable the buffering in `OutputStreamDataOutput`)

mikemccand commented on core/src/main/java/org/elasticsearch/index/translog/Checkpoint.java:
```diff
@@ -18,86 +18,135 @@
-        final ByteArrayDataOutput out = new ByteArrayDataOutput(buffer);
-        write(out);
-        Channels.writeToChannel(buffer, channel);
+    static Checkpoint readChecksummed(DataInput in) throws IOException {
```
Maybe rename to `readChecksummedV1`?

bleskes replied:
will do.

mikemccand commented:
LGTM, I just left a minor renaming suggestion.

s1monw commented:
LGTM 2
